File: vivaldi/test/vector_test.exs
defmodule VectorTest do
  use ExUnit.Case
  
  alias Vivaldi.Simulation.Vector

  test "Vector Distance" do
    points_and_results = [
     {[0, 0], [0, 0], 0},
     {[0, 0], [0, 1], 1},
     {[0, 0], [0, 2], 2},
     {[0, 3], [4, 0], 5},
     {[0, 1], [1, 0], :math.sqrt(2)}
    ]

    Enum.each(points_and_results, fn {p1, p2, expected_distance} ->
      result = Vector.distance(p1, p2)
      assert abs(result - expected_distance) < 0.00001
    end)
  end

  test "Vector Diff" do
    points_and_results = [
     {[0, 0], [0, 0], [0, 0]},
     {[0, 1], [0, 0], [0, 1]},
     {[2, 1], [1, 2], [1, -1]},
     {[0, 3], [4, 0], [-4, 3]},
    ]

    Enum.each(points_and_results, fn {p1, p2, expected} ->
      result = Vector.diff(p1, p2)
      assert expected == result
    end)
  end

  test "Vector Magnitude" do
    points_and_results = [
     {[0, 0], 0},
     {[0, 3, 4], 5},
     {[1, 1], :math.sqrt(2)},
     {[-1, -1], :math.sqrt(2)},
    ]

    Enum.each(points_and_results, fn {p1, expected} ->
      result = Vector.magnitude(p1)
      assert expected == result
    end)
  end

  test "Vector Scale" do
    inputs_and_results = [
     {[0, 0], 100, [0, 0]},
     {[0, 3, 4], 2, [0, 6, 8]},
     {[1, 1], -4, [-4, -4]},
     {[-1, -1], -4, [4, 4]},
    ]

    Enum.each(inputs_and_results, fn {p1, factor, expected} ->
      result = Vector.scale(p1, factor)
      assert expected == result
    end)
  end

  test "Unit vector at" do
    inputs_and_results = [
     {[1, 1], [0, 0], [:math.cos(:math.pi/4), :math.sin(:math.pi/4)]},
     {[2, 1], [2, 0], [0, 1]}
    ]

    Enum.each(inputs_and_results, fn {p1, p2, expected} ->
      result = Vector.unit_vector_at(p1, p2)
      Enum.zip(result, expected)
      |> Enum.map(fn {r, e} ->
        assert abs(r - e) < 0.0001
      end)
    end)

    rand_unit_vector = Vector.unit_vector_at([0, 0], [0, 0])
    assert abs(Vector.magnitude(rand_unit_vector) - 1.0) < 0.00001
  end
end

File: visualization/__tests__/vector-test.js
import Vector from './../src/vector';

// all binary vector operations are assumed to be called with matching dimensions.

describe('add', () => {
  it('should add two vectors ', () => {
    const pointsAndResults = [
       [[1], [2], [3]],
       [[0, 0], [0, 0], [0, 0]],
       [[0, 1], [0, 0], [0, 1]],
       [[2, 1], [1, 2], [3, 3]],
       [[0, 3], [4, 0], [4, 3]],
    ];


    pointsAndResults.map((v) => {
      expect(Vector.add(v[0], v[1])).toEqual(v[2])
    })
  });
});

describe('diff', () => {
  it('should add diff vectors', () => {
    const pointsAndResults = [
      [[1], [2], [-1]],
      [[0, 0], [0, 0], [0, 0]],
      [[0, 1], [0, 0], [0, 1]],
      [[2, 1], [1, 2], [1, -1]],
      [[0, 3], [4, 0], [-4, 3]],
    ];

    pointsAndResults.map((v) => {
      expect(Vector.diff(v[0], v[1])).toEqual(v[2])
    })
  });
});

describe('distance', () => {
  it('should calculate distance between two vectors', () => {
    const pointsAndResults = [
      [[1], [1], 0.0],
      [[0, 0], [0, 0], 0.0],
      [[0, 1], [0, 0], 1.0],
      [[0, 0], [2, 0], 2.0],
      [[2, 1], [1, 2], Math.sqrt(2)],
      [[0, 3], [4, 0], 5.0],
    ];

    pointsAndResults.map((v) => {
      expect(Vector.distance(v[0], v[1])).toBe(v[2])
    })
  });
});

describe('magnitude', () => {
  it('should calculate distance of given vector from origin', () => {
    const pointAndResult = [
      [[1], 1.0],
      [[0, 0], 0.0],
      [[0, 1], 1.0],
      [[2, 0], 2.0],
      [[1, 1], Math.sqrt(2)],
      [[3, 4], 5.0],
    ];

    pointAndResult.map((v) => {
      expect(Vector.magnitude(v[0])).toBe(v[1])
    })
  });
});

describe('scale', () => {
  it('should scale vector by given factor', () => {
    const pointFactorAndResult = [
      [[1], 0, [0]],
      [[2], 1, [2]],
      [[0, 0], 1, [0, 0]],
      [[0, 1], 2, [0, 2]],
      [[1, 2], -2, [-2, -4]],
    ];

    pointFactorAndResult.map((v) => {
      expect(Vector.scale(v[0], v[1])).toEqual(v[2])
    })
  });
});


describe('unitVectorAt', () => {
  // ```
  //   Given two vectors p1 and p2, 
  //   if p1 and p2 are both equal to zero,
  //     it should return a random unit vector,
  //   else
  //     it should return unitVectorAt p1 - p2
  // ```
  it('should return a unit vector', () => {

    expect.extend({
      toBeARandomUnitVector(received) {
        const pass = (Vector.magnitude(received) - 1.0) < 0.0001
        return {
          pass: pass,
          message: () => `expected ${received} to be a random unit vector`,
        }
      },

      toBeApproximatelyEqualVectors(a, b) {
        const pass = (function() {
          for (let i = 0, length = a.length; i < length; i++) {
            if (Math.abs(a[i] - b[i]) > 0.0001) {
              return false
            }
          }
          return true
        })()

        return {
          pass: pass,
          message: () => `expected ${b} to be a approximately equal to ${a}`,
        }
      }
    });

    expect(Vector.unitVectorAt([0, 0], [0, 0])).toBeARandomUnitVector()

    const pointsAndResults = [
      [[2, 2], [1, 1], [Math.cos(Math.PI/4), Math.sin(Math.PI/4)]],
      [[2, 1], [2, 0], [0, 1]],
    ];

    pointsAndResults.map((v) => {
      expect(Vector.unitVectorAt(v[0], v[1])).toBeApproximatelyEqualVectors(v[2])
    })

  });
});
